### PR TITLE
Awikstats stats module update

### PR DESCRIPTION
**Pull request by avyayv**
See [this](https://github.com/apertium/phenny/issues/389) issue

### CODE REVIEW

File: apertium_wikistats_bot.py [hunk 1]
```diff
@@ -0,0 +1,520 @@
...
+
+import argparse, requests, json, logging, sys, re, os, subprocess, shutil, importlib, urllib.request, collections, tempfile
+import xml.etree.ElementTree as etree
+import datetime
+import autocoverage
+
+s = requests.Session()
+
+def countStems(dictionary, vanilla=False):
+    url = "http://apertium.projectjj.com/stats-service/apertium-"+dictionary
```
Comment (sushain97) [inline apertium_wikistats_bot.py]: Please don't hardcode this everywhere. Use a constant at the top. Furthermore, don't concatenate strings, use format strings.

File: apertium_wikistats_bot.py [hunk 2]
```diff
@@ -0,0 +1,520 @@
...
+    statsServiceResult = s.get(url)
```
Comment (sushain97) [inline apertium_wikistats_bot.py]:
```suggestion
    statsServiceResult = s.get(url).json()
```
probably works

File: apertium_wikistats_bot.py [hunk 3]
```diff
@@ -0,0 +1,520 @@
...
+            if stat["stat_kind"] == "Stems":
+                return stat["value"]
+    else:
+        for stat in statsServiceJsonResult["stats"]:
+            if stat["stat_kind"] == "VanillaStems":
+                return stat["value"]
+    return
+
+def queryForValue(lang, file_kind, stat_kind=None):
+    if(stat_kind):
```
Comment (sushain97) [inline apertium_wikistats_bot.py]: Don't use parens in Python conditionals.

File: apertium_wikistats_bot.py (hunk 2 above)
Comment (sushain97) [reply apertium_wikistats_bot.py]: Here and everywhere else.

File: apertium_wikistats_bot.py [hunk 4]
```diff
@@ -0,0 +1,520 @@
...
+        statsServiceJsonResult = json.loads(statsServiceResult.text)
+        if "stats" in statsServiceJsonResult:
+            for stat in statsServiceJsonResult["stats"]:
+                if stat["file_kind"] == file_kind and stat["stat_kind"] == stat_kind:
+                    return stat["value"]
+
+def getCounts(lang, fileFormat):
+    if fileFormat == 'Monodix':
+        return {
+            'stems': queryForValue(lang, fileFormat, "Stems"),
```
Comment (sushain97) [inline apertium_wikistats_bot.py]: You're making multiple queries every single time instead of just making one query and getting all the info from it.

File: bot.py [hunk 5]
```diff
@@ -52,49 +52,49 @@ def module_control(phenny, module, func):
...
                 filenames.append(os.path.join(home, 'modules', fn + '.py'))
 
-        if hasattr(self.config, 'extra'): 
-            for fn in self.config.extra: 
-                if os.path.isfile(fn): 
+        if hasattr(self.config, 'extra'):
+            for fn in self.config.extra:
+                if os.path.isfile(fn):
                     filenames.append(fn)
-                elif os.path.isdir(fn):
```
Comment (sushain97) [inline bot.py]: Can you send a different PR with all these whitespace changes? They make this diff really hard to review.

File: apertium_wikistats_bot.py [hunk 6]
```diff
@@ -0,0 +1,520 @@
...
+                    if revisionInfo:
+                        fileCounts[lang + ' ' + countType] = (count, revisionInfo, "https://raw.githubusercontent.com/apertium/"+pair+"/master/"+fileLoc)
+    return fileCounts
+
+def getMonoLangCounts(lang):
+    fileCounts = {}
+
+    url = "http://apertium.projectjj.com/stats-service/apertium-"+lang
+    statsServiceResult = s.get(url)
+    statsServiceJsonResult = json.loads(statsServiceResult.text)
```
Comment (sushain97) [inline apertium_wikistats_bot.py]: There really should only be one place where code like this occurs.

File: apertium_wikistats_bot.py [hunk 7]
```diff
@@ -0,0 +1,520 @@
...
+
+    if getCounts(lang, 'Monodix'):
+        if "stats" in statsServiceJsonResult:
+            for stat in statsServiceJsonResult["stats"]:
+                if stat["file_kind"] == 'Monodix':
+                    type = stat["file_kind"]
```
Comment (sushain97) [inline apertium_wikistats_bot.py]: Don't use variables that shadow a builtin., in this case, `type`.

File: apertium_wikistats_bot.py [hunk 8]
```diff
@@ -0,0 +1,520 @@
...
+        for countType, count in counts.items():
+            revisionInfo = getRevisionInfo(lang, type)
+            if revisionInfo:
+                fileCounts[countType] = (count, revisionInfo, "https://raw.githubusercontent.com/apertium/apertium-"+lang+"/master/"+fileLoc)
+
+    return fileCounts
+
+def getRevisionInfo(lang, file_kind):
+    try:
+        url = "http://apertium.projectjj.com/stats-service/apertium-"+lang
```
Comment (sushain97) [inline apertium_wikistats_bot.py]: Why are you requesting all the info again..?

File: apertium_wikistats_bot.py [hunk 9]
```diff
@@ -0,0 +1,506 @@
...
+            'rlx rules': queryForValue(statsServiceJsonResult, fileFormat, "Rules")
+        }
+
+    return
+
+def countAllStats(statsServiceJsonResult, arr):
+    fileCounts = {}
+    if "stats" in statsServiceJsonResult:
+        for stat in statsServiceJsonResult["stats"]:
+
```
Comment (sushain97) [inline apertium_wikistats_bot.py]:
```suggestion
```

File: apertium_wikistats_bot.py [hunk 10]
```diff
@@ -0,0 +1,506 @@
...
+            if format == "MetaMonodix":
+                lang_pair_to_post = toISO(fileLoc.split('.')[1].split('.')[0])
+            if counts:
+                for countType, count in counts.items():
+                    revisionInfo = getRevisionInfo(statsServiceJsonResult, format)
+                    if revisionInfo:
+                        fileCounts[lang_pair_to_post + ' ' + countType] = (count, revisionInfo, githubBlobUrl % (pair, fileLoc))
+    return fileCounts
+
+def getJSONFromStatsService(lang):
```
Comment (sushain97) [inline apertium_wikistats_bot.py]: Instead of `GET`, use `POST` since that will ensure updated statistics are provided. You'll need to provide the `async=false` param so that you don't have to poll.

File: apertium_wikistats_bot.py [hunk 11]
```diff
@@ -0,0 +1,519 @@
...
+import argparse, requests, json, logging, sys, re, os, subprocess, shutil, importlib, urllib.request, collections, tempfile
+import xml.etree.ElementTree as etree
+import datetime
+import autocoverage
+
+s = requests.Session()
+
+def countStems(statsServiceJsonResult, vanilla=False):
+    if vanilla == False:
+        for rawStats in statsServiceJsonResult:
```
Comment (sushain97) [inline apertium_wikistats_bot.py]: Er, I meant `statsServiceJsonResult` -> `rawStats`. The former is a mouthful.

File: apertium_wikistats_bot.py [hunk 12]
```diff
@@ -0,0 +1,519 @@
...
+            if rawStats['stat_kind'] == 'Stems':
+                return rawStats['value']
+    else:
+        for rawStats in statsServiceJsonResult:
+            if rawStats['stat_kind'] == 'VanillaStems':
+                return rawStats['value']
+    return
```
Comment (sushain97) [inline apertium_wikistats_bot.py]:
```suggestion
```

File: apertium_wikistats_bot.py [hunk 13]
```diff
@@ -0,0 +1,519 @@
...
+        return {
+            'stems': queryForValue(statsServiceJsonResult, fileFormat, 'Stems')
+        }
+    elif fileFormat == 'MetaBidix':
+        return {
+            'meta stems': queryForValue(statsServiceJsonResult, fileFormat, 'Stems')
+        }
+    elif fileFormat == 'Lexc':
+        return {
+            'stems': countStems(statsServiceJsonResult),
```
Comment (sushain97) [inline apertium_wikistats_bot.py]: Is there a reason we can't just use `queryForValue` here?

File: apertium_wikistats_bot.py [hunk 14]
```diff
@@ -0,0 +1,519 @@
...
+
+    return
+
+def countAllStats(statsServiceJsonResult, arr):
+    fileCounts = {}
+    for rawStats in statsServiceJsonResult:
+        format = rawStats['file_kind']
+        fileLoc = rawStats['path']
+        pair = rawStats['name']
+        lang_pair_to_post = pair.split('apertium-')[1]
```
Comment (sushain97) [inline apertium_wikistats_bot.py]: Why are we swapping between `camelCase` and `snake_case`?

File: apertium_wikistats_bot.py [hunk 15]
```diff
@@ -0,0 +1,519 @@
...
+            pair_for_url = isoCodeLangPair
+        url = statsURL % pair_for_url
+        statsServiceJsonResult = s.post(url).json()
+    else:
+        if toAlpha3Code(pair_for_url):
+            isoCodeLangPair = toAlpha3Code(pair_for_url)
+        if isoCodeLangPair != "":
+            pair_for_url = isoCodeLangPair
+        url = statsURL % pair_for_url
+        statsServiceJsonResult = s.post(url).json()
```
Comment (sushain97) [inline apertium_wikistats_bot.py]: Can this line just be pulled out of the conditionals? Actually, these last two.

File: apertium_wikistats_bot.py [hunk 16]
```diff
@@ -0,0 +1,519 @@
...
+    if 'stats' in statsServiceJsonResult:
+        return statsServiceJsonResult['stats']
+    logging.error('Unable to find %s' % isoCodeLangPair)
+    return
```
Comment (sushain97) [inline apertium_wikistats_bot.py]:
```suggestion
```

File: apertium_wikistats_bot.py [hunk 17]
```diff
@@ -0,0 +1,519 @@
...
+            revisionInfo = getRevisionInfo(statsServiceJsonResult, file_kind)
+            if revisionInfo:
+                fileCounts[countType] = (count, revisionInfo, githubBlobUrl % (statsServiceJsonResult, fileLoc))
+    if getCounts(statsServiceJsonResult, 'Lexc'):
+        for rawStats in statsServiceJsonResult:
+            if rawStats['file_kind'] == 'Monodix':
+                file_kind = rawStats['file_kind']
+                fileLoc = rawStats['path']
+        counts = getCounts(statsServiceJsonResult, 'Lexc')
+        for countType, count in counts.items():
```
Comment (sushain97) [inline apertium_wikistats_bot.py]: This block seems repeated in each case?

File: apertium_wikistats_bot.py [hunk 18]
```diff
@@ -0,0 +1,519 @@
...
+    return fileCounts
+
+def getRevisionInfo(statsServiceJsonResult, file_kind):
+    try:
+        for rawStats in statsServiceJsonResult:
+            if rawStats['file_kind'] == file_kind:
+                revisionNumber = rawStats['revision']
+                revisionAuthor = rawStats['last_author']
+                return (revisionNumber, revisionAuthor)
+        return
```
Comment (sushain97) [inline apertium_wikistats_bot.py]:
```suggestion
```

File: apertium_wikistats_bot.py [hunk 19]
```diff
@@ -0,0 +1,519 @@
...
+    except Exception as e:
+        logging.error('Unable to get revision info for %s: %s' % (uri, e))
+        return None
```
Comment (sushain97) [inline apertium_wikistats_bot.py]:
```suggestion
```

File: apertium_wikistats_bot.py [hunk 20]
```diff
@@ -0,0 +1,519 @@
+#!/usr/bin/env python3
+
+wikiURL = 'http://wiki.apertium.org/wiki/'
+apiURL = 'http://wiki.apertium.org/w/api.php'
+statsURL = 'http://apertium.projectjj.com/stats-service/apertium-%s/?async=false'
+githubBlobUrl = 'https://raw.githubusercontent.com/apertium/%s/master/%s'
+import argparse, requests, json, logging, sys, re, os, subprocess, shutil, importlib, urllib.request, collections, tempfile
```
Comment (sushain97) [inline apertium_wikistats_bot.py]:
```suggestion

import argparse, requests, json, logging, sys, re, os, subprocess, shutil, importlib, urllib.request, collections, tempfile
```

File: apertium_wikistats_bot.py [hunk 21]
```diff
@@ -0,0 +1,519 @@
+#!/usr/bin/env python3
+
+wikiURL = 'http://wiki.apertium.org/wiki/'
```
Comment (sushain97) [inline apertium_wikistats_bot.py]: Move these below the imports.

File: apertium_wikistats_bot.py [hunk 22]
```diff
@@ -0,0 +1,489 @@
...
+import autocoverage
+wikiURL = 'http://wiki.apertium.org/wiki/'
+apiURL = 'http://wiki.apertium.org/w/api.php'
+statsURL = 'http://apertium.projectjj.com/stats-service/apertium-%s/?async=false'
+githubBlobUrl = 'https://raw.githubusercontent.com/apertium/%s/master/%s'
+
+s = requests.Session()
+
+def queryForValue(rawStats, fileKind, stat_kind):
+    if stat_kind:
```
Comment (sushain97) [inline apertium_wikistats_bot.py]: Why would you call this function with `stat_kind is None`?

File: apertium_wikistats_bot.py [hunk 23]
```diff
@@ -0,0 +1,489 @@
...
+import datetime
+import autocoverage
+wikiURL = 'http://wiki.apertium.org/wiki/'
+apiURL = 'http://wiki.apertium.org/w/api.php'
+statsURL = 'http://apertium.projectjj.com/stats-service/apertium-%s/?async=false'
+githubBlobUrl = 'https://raw.githubusercontent.com/apertium/%s/master/%s'
+
+s = requests.Session()
+
+def queryForValue(rawStats, fileKind, stat_kind):
```
Comment (sushain97) [inline apertium_wikistats_bot.py]: Both camelCase and snake_case again...

File: apertium_wikistats_bot.py [hunk 24]
```diff
@@ -0,0 +1,489 @@
...
+
+def countAllStats(rawStats, arr):
+    fileCounts = {}
+    for stat in rawStats:
+        format = stat['file_kind']
+        fileLoc = stat['path']
+        pair = stat['name']
+        lang_pair_to_post = pair.split('apertium-')[1]
+        counts = getCounts(rawStats, format)
+        if format == 'MetaMonodix':
```
Comment (sushain97) [inline apertium_wikistats_bot.py]: What is this special case for?

File: apertium_wikistats_bot.py [hunk 25]
```diff
@@ -0,0 +1,489 @@
...
+    if toAlpha3Code(pairForUrl):
+        isoCodeLangPair = toAlpha3Code(pairForUrl)
+    if len(indLangs) == 2:
+        if toAlpha3Code(indLangs[0]):
+            isoCodeLangPair = toAlpha3Code(indLangs[0]) + '-'
+        if toAlpha3Code(indLangs[1]):
+            isoCodeLangPair += toAlpha3Code(indLangs[1])
+        if isoCodeLangPair != "":
+            pairForUrl = isoCodeLangPair
+        url = statsURL % pairForUrl
```
Comment (sushain97) [inline apertium_wikistats_bot.py]: Can't these two lines still be moved out of the conditionals?

File: apertium_wikistats_bot.py [hunk 26]
```diff
@@ -0,0 +1,489 @@
+#!/usr/bin/env python3
+
+import argparse, requests, json, logging, sys, re, os, subprocess, shutil, importlib, urllib.request, collections, tempfile
+
```
Comment (sushain97) [inline apertium_wikistats_bot.py]:
```suggestion
```

File: apertium_wikistats_bot.py [hunk 27]
```diff
@@ -0,0 +1,489 @@
...
+import xml.etree.ElementTree as etree
+import datetime
+import autocoverage
```
Comment (sushain97) [inline apertium_wikistats_bot.py]:
```suggestion
import autocoverage

```

File: apertium_wikistats_bot.py [hunk 28]
```diff
@@ -0,0 +1,489 @@
...
+s = requests.Session()
+
+def queryForValue(rawStats, fileKind, stat_kind):
+    if stat_kind:
+        for stat in rawStats:
+            if stat['file_kind'] == fileKind and stat['stat_kind'] == stat_kind:
+                return stat['value']
+
+def getCounts(rawStats, fileFormat):
+    if fileFormat == 'Monodix':
```
Comment (sushain97) [inline apertium_wikistats_bot.py]: Seems like this whole block of stuff could be a dictionary from fileFormat to dictionary of key => rawStats key.

e.g. `{'Monodix' => {'stems' => 'Stems', paradigms => 'Paradigms'}, ...}`

Then, the code itself would just be a simple loop.

File: apertium_wikistats_bot.py [hunk 29]
```diff
@@ -0,0 +1,489 @@
+#!/usr/bin/env python3
+
+import argparse, requests, json, logging, sys, re, os, subprocess, shutil, importlib, urllib.request, collections, tempfile
```
Comment (sushain97) [inline apertium_wikistats_bot.py]: Could we just do 1 per line? I think this isn't on you since the code used to be this way but it's super inconsistent now.

**Comment by avyayv** [comment]
**.awikstats update cat**
```
15:51 <freenode> phennyINFO:root:Login as avyayv succeeded
15:51 <freenode> phennyINFO:root:Acquired file counts {'monodix stems': (95551, (1449, 'jaumeortola'), 'https://raw.githubusercontent.com/apertium/apertium-cat/master/apertium-cat.cat.dix'), 'paradigms': (683, (1449, 'jaumeortola'), 'https://raw.githubusercontent.com/apertium/apertium-cat/master/apertium-cat.cat.dix'), 'lexc stems': (None, (1449, 'jaumeortola'), 'https://raw.githubusercontent.com/apertium/apertium-cat/master/apertium-cat.cat.dix'),
15:51 <freenode> phenny'vanilla stems': (None, (1449, 'jaumeortola'), 'https://raw.githubusercontent.com/apertium/apertium-cat/master/apertium-cat.cat.dix'), 'rlx rules': ('596', (1449, 'jaumeortola'), 'https://raw.githubusercontent.com/apertium/apertium-cat/master/apertium-cat.cat.dix')}
15:52 <freenode> phennyINFO:root:Update of page Apertium-cat/stats succeeded (http://wiki.apertium.org/wiki/Apertium-cat/stats)
```
**.awikstats update kaz-tat**
```
16:26 <freenode> phennyINFO:root:Update of page Apertium-kaz-tat/stats succeeded (http://wiki.apertium.org/wiki/Apertium-kaz-tat/stats)
16:26 <freenode> phennyINFO:root:Login as avyayv succeeded
```
**.awikstats update en-es**
`16:36 <freenode> phennyINFO:root:Update of page Apertium-en-es/stats succeeded (http://wiki.apertium.org/wiki/Apertium-en-es/stats)`
**.awikstats update kaz**
```
16:52 <freenode> phennyINFO:root:Login as avyayv succeeded
16:52 <freenode> phennyINFO:root:Acquired file counts {'lexc stems': (36161, (4060, 'piraye'), 'https://raw.githubusercontent.com/apertium/apertium-kaz/master/apertium-kaz.kaz.lexc'), 'vanilla stems': (26984, (4060, 'piraye'), 'https://raw.githubusercontent.com/apertium/apertium-kaz/master/apertium-kaz.kaz.lexc'), 'rlx rules': ('143', (3952, 'ilnar.salimzianov'), 'https://raw.githubusercontent.com/apertium/apertium-kaz/master/apertium-kaz.kaz.rlx')}
16:52 <freenode> phennyINFO:root:Update of page Apertium-kaz/stats succeeded (http://wiki.apertium.org/wiki/Apertium-kaz/stats)
```

**Comment by sushain97** [comment]
Pretty sure it didn't work. There are a bunch of lines with `None`. e.g.:

> kaz-tat stems: None as of r1258 by jonathan.north.washington ~ Avyayv (talk) 01:26, 3 December 2018 (CET), run by avyayv

File: apertium_wikistats_bot.py [hunk 30]
```diff
@@ -0,0 +1,502 @@
...
+                revisionInfo = getRevisionInfo(rawStats, format)
+                if revisionInfo:
+                    fileCounts[countType] = (count, revisionInfo, githubBlobUrl % (pair, fileLoc))
+    return fileCounts
+
+def getJSONFromStatsService(lang):
+    pairForUrl = lang
+    isoCodeLangPair = ""
+    indLangs = pairForUrl.split('-')
+    if toAlpha3Code(pairForUrl):
```
Comment (sushain97) [inline apertium_wikistats_bot.py]: The logic from L81-89 is quite convoluted and I'm having a hard time following it. Instead, why can't you `map` `toAlpha3Code` over the `indLangs` and then join with `-`?

File: apertium_wikistats_bot.py [hunk 31]
```diff
@@ -0,0 +1,502 @@
...
+            isoCodeLangPair = toAlpha3Code(indLangs[0]) + '-'
+        if toAlpha3Code(indLangs[1]):
+            isoCodeLangPair += toAlpha3Code(indLangs[1])
+    if isoCodeLangPair != "":
+        pairForUrl = isoCodeLangPair
+    url = statsURL % pairForUrl
+    rawStats = s.post(url).json()
+    if 'stats' in rawStats:
+        return rawStats['stats']
+    logging.error('Unable to find %s' % isoCodeLangPair)
```
Comment (sushain97) [inline apertium_wikistats_bot.py]:
```suggestion
    logging.error('Unable to request stats for %s' % isoCodeLangPair)
```

File: apertium_wikistats_bot.py [hunk 32]
```diff
@@ -0,0 +1,502 @@
...
+            if stat['file_kind'] == typeOfDict:
+                fileKind = stat['file_kind']
+                fileLoc = stat['path']
+                fileName = stat['name']
+        counts = getCounts(rawStats, typeOfDict)
+        for countType, count in counts.items():
+            revisionInfo = getRevisionInfo(rawStats, fileKind)
+            if revisionInfo:
+                fileCounts[countType] = (count, revisionInfo, githubBlobUrl % (fileName, fileLoc))
+    except UnboundLocalError:
```
Comment (sushain97) [inline apertium_wikistats_bot.py]: What...? This is a really bad coding practice since it'll shadow real errors.

File: apertium_wikistats_bot.py [hunk 33]
```diff
@@ -0,0 +1,502 @@
...
+        pass
+
+def getMonoLangCounts(rawStats):
+    fileCounts = {}
+    if getCounts(rawStats, 'Monodix'):
```
Comment (sushain97) [inline apertium_wikistats_bot.py]: Seems like these 6 lines can be replaced with a loop?

File: apertium_wikistats_bot.py [hunk 34]
```diff
@@ -0,0 +1,502 @@
...
+            isoCodeLangPair += toAlpha3Code(indLangs[1])
+    if isoCodeLangPair != "":
+        pairForUrl = isoCodeLangPair
+    url = statsURL % pairForUrl
+    rawStats = s.post(url).json()
+    if 'stats' in rawStats:
+        return rawStats['stats']
+    logging.error('Unable to find %s' % isoCodeLangPair)
+
+def monoLangInformation(typeOfDict, rawStats, fileCounts):
```
Comment (sushain97) [inline apertium_wikistats_bot.py]:
```suggestion
def monoLangInformation(fileFormat, rawStats, fileCounts):
```

File: apertium_wikistats_bot.py [hunk 35]
```diff
@@ -0,0 +1,502 @@
...
+
+        if templateLine.match(line):
+            if 'Corpora' in line:
+                isCorpora = True
+        elif line.startswith(wpName):
+            isAfter = True
+        elif isCorpora and isAfter and line == '':
+            isCorpora = False
+
+    words, coverage, revision_num = human_format(words), '{0:.1f}'.format(coverage), getRevisionInfo('https://svn.code.sf.net/p/apertium/svn/languages/apertium-' + lang + '/')
```
Comment (sushain97) [inline apertium_wikistats_bot.py]: This line references SVN?

File: apertium_wikistats_bot.py [hunk 36]
```diff
@@ -0,0 +1,502 @@
...
+                            count[langPairToPost + ' ' + wikiKey] = queryForValue(rawStats, fileFormat, countDict[fileFormat][key], langPairToPost, fileLoc)
+                    count[langPairToPost + ' ' + wikiKey] = queryForValue(rawStats, fileFormat, countDict[fileFormat][key], langPairToPost, fileLoc)
+                else:
+                    count[wikiKey] = queryForValue(rawStats, fileFormat, countDict[fileFormat][key])
+        return count
+
+def countAllStats(rawStats, arr):
+    fileCounts = {}
+    for stat in rawStats:
+        format = stat['file_kind']
```
Comment (sushain97) [inline apertium_wikistats_bot.py]:
```suggestion
        fileFormat = stat['file_kind']
```

File: apertium_wikistats_bot.py (hunk 18 above)
Comment (sushain97) [reply apertium_wikistats_bot.py]: This was not resolved.

File: apertium_wikistats_bot.py [hunk 37]
```diff
@@ -0,0 +1,502 @@
...
+    if getCounts(rawStats, 'Lexc'):
+        monoLangInformation('Lexc', rawStats, fileCounts)
+    if getCounts(rawStats, 'Rlx'):
+        monoLangInformation('Rlx', rawStats, fileCounts)
+    return fileCounts
+
+def getRevisionInfo(rawStats, fileKind):
+    try:
+        for stat in rawStats:
+            if stat['file_kind'] == fileKind:
```
Comment (sushain97) [inline apertium_wikistats_bot.py]: What if there are multiple of the same `fileKind`, as there often are? This logic seems flawed and it seems like an extra loop is entirely unnecessary. The stat value, revision and author can be probably be passed around together (or just pass the entry in the `stats` array itself).

File: apertium_wikistats_bot.py [hunk 38]
```diff
@@ -0,0 +1,507 @@
...
+import autocoverage
+
+wikiURL = 'http://wiki.apertium.org/wiki/'
+apiURL = 'http://wiki.apertium.org/w/api.php'
+statsURL = 'http://apertium.projectjj.com/stats-service/apertium-%s/?async=false'
+githubBlobUrl = 'https://raw.githubusercontent.com/apertium/%s/master/%s'
+
+s = requests.Session()
+
+fileStatTypeMapping = {'Monodix': {'stems': 'Stems', 'paradigms': 'Paradigms'},
```
Comment (sushain97) [inline apertium_wikistats_bot.py]:
```suggestion
fileStatTypeMapping = {
    'Monodix': {'stems': 'Stems', 'paradigms': 'Paradigms'},
```

File: apertium_wikistats_bot.py [hunk 39]
```diff
@@ -0,0 +1,507 @@
...
+    'MetaMonodix': {'meta stems': 'Stems', 'meta paradigms': 'Paradigms'},
+    'Bidix': {'stems': 'Entries'},
+    'MetaBidix': {'meta stems': 'Stems'},
+    'Lexc': {'stems': 'Stems', 'vanilla stems': 'VanillaStems'},
+    'Rlx': {'rlx rules': 'Rules'},
+    'Transfer': {'rules': 'Rules', 'macros': 'Macros'}
+    }
```
Comment (sushain97) [inline apertium_wikistats_bot.py]:
```suggestion
}
```

File: apertium_wikistats_bot.py [hunk 40]
```diff
@@ -0,0 +1,507 @@
...
+
+def getStatValue(rawStats, fileKind, statKind, pair=None, fileLoc=None):
+    for stat in rawStats:
+        if pair and fileLoc:
+            langAndExtension = stat['path'].split('.')
+            pairInFile = langAndExtension[1]
+            extension = langAndExtension[-1]
+            extensionMatches = (extension == fileLoc.split('.')[-1])
```
Comment (sushain97) [inline apertium_wikistats_bot.py]:
```suggestion
            extensionMatches = extension == fileLoc.split('.')[-1]
```

File: apertium_wikistats_bot.py [hunk 41]
```diff
@@ -0,0 +1,507 @@
...
+            pairMatches = (pair == pairInFile)
```
Comment (sushain97) [inline apertium_wikistats_bot.py]:
```suggestion
            pairMatches = pair == pairInFile
```

Comment (sushain97) [reply apertium_wikistats_bot.py]: Honestly, I would just make the comparison below span multiple lines. This construction is even more difficult to read.

File: apertium_wikistats_bot.py [hunk 42]
```diff
@@ -0,0 +1,507 @@
...
+            if getStatValue(rawStats, fileFormat, fileStatTypeMapping[fileFormat][key]) != None:
+                wikiKey = key
+                if langPairToPost and fileLoc:
+                    for stat in rawStats:
+                        langAndExtension = stat['path'].split('.')
+                        pairInFile = langAndExtension[1]
+                        extension = langAndExtension[-1]
+                        extensionAndPairEqual = (langPairToPost == pairInFile and extension == fileLoc.split('.')[-1])
+                        statKindsEqual = (stat['stat_kind'] == fileStatTypeMapping[fileFormat][key])
+                        if stat['file_kind'] == 'Transfer' and statKindsEqual and extensionAndPairEqual:
```
Comment (sushain97) [inline apertium_wikistats_bot.py]: Same comment here re. comparison spanning multiple lines.

File: apertium_wikistats_bot.py [hunk 43]
```diff
@@ -0,0 +1,507 @@
...
+                return stat['value']
+        else:
+            if stat['file_kind'] == fileKind and stat['stat_kind'] == statKind:
+                return stat['value']
+
+def getCounts(rawStats, fileFormat, langPairToPost=None, fileLoc=None):
+    count = {}
+    if fileFormat in fileStatTypeMapping:
+        for key in fileStatTypeMapping[fileFormat]:
+            if getStatValue(rawStats, fileFormat, fileStatTypeMapping[fileFormat][key]) != None:
```
Comment (sushain97) [inline apertium_wikistats_bot.py]:
```suggestion
            if getStatValue(rawStats, fileFormat, fileStatTypeMapping[fileFormat][key]) is not None:
```

File: apertium_wikistats_bot.py [hunk 44]
```diff
@@ -0,0 +1,507 @@
...
+        indLangs = pairForUrl.split('-')
+        isoCodeLangPair = '-'.join(list(map(toAlpha3Code, indLangs)))
+        if isoCodeLangPair:
+            pairForUrl = isoCodeLangPair
+        url = statsURL % pairForUrl
+        rawStats = s.post(url).json()
+        if 'stats' in rawStats:
+            return rawStats['stats']
+        raise ValueError('Cannot find this module')
+    except ValueError:
```
Comment (sushain97) [inline apertium_wikistats_bot.py]: What's with the mental gymnastics of raising and immediately catching an exception? Just do what you want to do.

File: apertium_wikistats_bot.py [hunk 45]
```diff
@@ -0,0 +1,507 @@
...
+                        fileCounts[countType] = (count, revisionInfo, githubBlobUrl % (pair, fileLoc))
+    return fileCounts
+
+def getJSONFromStatsService(lang):
+    try:
+        pairForUrl = lang
+        isoCodeLangPair = None
+        indLangs = pairForUrl.split('-')
+        isoCodeLangPair = '-'.join(list(map(toAlpha3Code, indLangs)))
+        if isoCodeLangPair:
```
Comment (sushain97) [inline apertium_wikistats_bot.py]: Why would this ever be false?

File: apertium_wikistats_bot.py [hunk 46]
```diff
@@ -0,0 +1,507 @@
...
+    file_types = ['Monodix', 'Lexc', 'Rlx']
+    fileCounts = {}
+    for file_type in file_types:
+        if getCounts(rawStats, file_type):
+            monoLangInformation(file_type, rawStats, fileCounts)
+    return fileCounts
+
+def getRevisionInfo(lang):
+    try:
+        commitUrl = "https://api.github.com/repos/apertium/apertium-%/commits" % lang
```
Comment (sushain97) [inline apertium_wikistats_bot.py]: This doesn't work. Not every file is at the latest revision. You need to use the revisions from the stats service.

File: apertium_wikistats_bot.py [hunk 47]
```diff
@@ -0,0 +1,510 @@
...
+    for stat in rawStats:
+        if pair and fileLoc:
+            langAndExtension = stat['path'].split('.')
+            pairInFile = langAndExtension[1]
+            extension = langAndExtension[-1]
+            extensionMatches = extension == fileLoc.split('.')[-1]
+            pairMatches = pair == pairInFile
+            if stat['file_kind'] == fileKind and \
+               stat['stat_kind'] == statKind and \
+               pairMatches and \
```
Comment (sushain97) [inline apertium_wikistats_bot.py]: This is worse than we started... I meant to stop using extra vars and just use a longer conditional broken into lines.

File: apertium_wikistats_bot.py [hunk 48]
```diff
@@ -0,0 +1,510 @@
...
+                wikiKey = key
+                if langPairToPost and fileLoc:
+                    for stat in rawStats:
+                        langAndExtension = stat['path'].split('.')
+                        pairInFile = langAndExtension[1]
+                        extension = langAndExtension[-1]
+                        pairEqual = langPairToPost == pairInFile
+                        extensionEqual = extension == fileLoc.split('.')[-1]
+                        statKindsEqual = (stat['stat_kind'] == fileStatTypeMapping[fileFormat][key])
+                        if stat['file_kind'] == 'Transfer' and \
```
Comment (sushain97) [inline apertium_wikistats_bot.py]: Why is transfer hardcoded here?

File: apertium_wikistats_bot.py [hunk 49]
```diff
@@ -0,0 +1,510 @@
...
+    fileCounts = {}
+    if rawStats:
+        for stat in rawStats:
+            format = stat['file_kind']
+            fileLoc = stat['path']
+            pair = stat['name']
+            lastAuthor = stat['last_author']
+            revisionNumber = stat['revision']
+            langPairToPost = stat['path'].split('.')[1]
+            counts = getCounts(rawStats, format, langPairToPost, fileLoc)
```
Comment (sushain97) [inline apertium_wikistats_bot.py]: Now that this code is shorter I can understand it a bit better and I'm really confused as to why you're iterating through stats in this function and then again in getCounts. Why do we need a N^2 loop?

**Comment by avyayv** [comment]
I removed all the extra loops so its easier to understand and is not as prone to errors.
**`.awikstats update kaz`**
```
15:46 <freenode> phennyINFO:root:Login as avyayv succeeded
15:46 <freenode> phennyINFO:root:Acquired file counts {'vanilla stems': (26984, (4060, 'piraye'), 'https://raw.githubusercontent.com/apertium/apertium-kaz/master/apertium-kaz.kaz.lexc'), 'stems': (36161, (4060, 'piraye'), 'https://raw.githubusercontent.com/apertium/apertium-kaz/master/apertium-kaz.kaz.lexc'), 'rlx rules': ('143', (3952, 'ilnar.salimzianov'), 'https://raw.githubusercontent.com/apertium/apertium-kaz/master/apertium-kaz.kaz.rlx')}
15:46 <freenode> phennyINFO:root:Update of page Apertium-kaz/stats succeeded (http://wiki.apertium.org/wiki/Apertium-kaz/stats)
```
**`.awikstats update en-es`**
```
15:54 <freenode> phennyINFO:root:Login as avyayv succeeded
15:54 <freenode> phennyINFO:root:Update of page Apertium-en-es/stats succeeded (http://wiki.apertium.org/wiki/Apertium-en-es/stats)
```
**`.awikstats update kaz-tat`**
```
15:55 <freenode> phennyINFO:root:Login as avyayv succeeded
15:55 <freenode> phennyINFO:root:Update of page Apertium-kaz-tat/stats succeeded (http://wiki.apertium.org/wiki/Apertium-kaz-tat/stats)
```
